x/twap/api.go

```diff
@@ -109,6 +120,134 @@ func (k Keeper) getTwap(
+	} else {
+		// If endTime < ctx.Height - interval
+		// We loop through each 2 days interval to find the accum of endTime
+		tempEndRecord := lastPreviousRecordKeep
+		for tempEndRecord.Time.Add(-interval).After(endTime) {
+			tracker, err := k.getAccumTracker(ctx, tempEndRecord.Time.Add(-interval), tempEndRecord, interval.Milliseconds())
+			if err != nil {
+				return sdk.Dec{}, err
+			}
+			tempEndRecord.Time = tracker.Time
+			tempEndRecord.P0ArithmeticTwapAccumulator = tracker.P0Accumulator
+			tempEndRecord.P1ArithmeticTwapAccumulator = tracker.P1Accumulator
+		}
+
+		diffTime := types.CanonicalTimeMs(tempEndRecord.Time) - types.CanonicalTimeMs(endTime)
+		if diffTime == 0 {
+			trackers = append(trackers, AccumTracker{P0Accumulator: tempEndRecord.P0ArithmeticTwapAccumulator, P1Accumulator: tempEndRecord.P1ArithmeticTwapAccumulator, Time: tempEndRecord.Time})
+		} else {
+			// With different time < interval, calculate accumulator directly from tempEndRecord
+			tracker, err := k.getAccumTracker(ctx, endTime, tempEndRecord, diffTime)
+			if err != nil {
+				return sdk.Dec{}, err
+			}
+			trackers = append(trackers, tracker)
+		}
+	}
+
+	// After have tracker, we find accum of startTime
+	for !trackers[0].Time.Add(-interval).Before(startTime) {
+		tempRecord := types.TwapRecord{
+			PoolId: lastPreviousRecordKeep.PoolId,
+			Asset0Denom: lastPreviousRecordKeep.Asset0Denom,
+			Asset1Denom: lastPreviousRecordKeep.Asset1Denom,
+			P0ArithmeticTwapAccumulator: trackers[0].P0Accumulator,
+			P1ArithmeticTwapAccumulator: trackers[0].P1Accumulator,
+		}
+		tracker, err := k.getAccumTracker(ctx, trackers[0].Time.Add(-interval), tempRecord, interval.Milliseconds())
+		if err != nil {
+			return sdk.Dec{}, err
+		}
+		trackers = append([]AccumTracker{tracker}, trackers...)
+	}
+
+	if !startTime.Equal(trackers[0].Time) {
+		tempRecord := types.TwapRecord{
+			PoolId: lastPreviousRecordKeep.PoolId,
+			Asset0Denom: lastPreviousRecordKeep.Asset0Denom,
+			Asset1Denom: lastPreviousRecordKeep.Asset1Denom,
+			P0ArithmeticTwapAccumulator: trackers[0].P0Accumulator,
+			P1ArithmeticTwapAccumulator: trackers[0].P1Accumulator,
+		}
+		tracker, err := k.getAccumTracker(ctx, startTime, tempRecord, trackers[0].Time.Sub(startTime).Milliseconds())
+		if err != nil {
+			return sdk.Dec{}, err
+		}
+		trackers = append([]AccumTracker{tracker}, trackers...)
+	}
+
+	timeDelta := types.CanonicalTimeMs(endTime) - types.CanonicalTimeMs(startTime)
+	var accumDiff sdk.Dec
+	if quoteAssetDenom == lastPreviousRecordKeep.Asset0Denom {
+		accumDiff = trackers[len(trackers) - 1].P0Accumulator.Sub(trackers[0].P0Accumulator)
+	} else {
+		accumDiff = trackers[len(trackers) - 1].P1Accumulator.Sub(trackers[0].P1Accumulator)
+	}
```

Comment:
If we're only using last element and first element, can we make it so that we use two different variables instead of having it as an array?